File: desktop/engine/models/agent.py
"""
Agent相关数据模型
"""
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Dict, Optional, Literal, Any
from pydantic import BaseModel, validator


# ============= 配置模型 =============

class ModelConfig(BaseModel):
    """模型配置（客户端版本 - 不包含敏感信息）"""
    provider: Literal["openai", "claude"]
    model: str
    temperature: float = 0.7
    max_tokens: int = 2000
    top_p: float = 1.0
    frequency_penalty: float = 0.0
    presence_penalty: float = 0.0
    # API Key 和 base_url 由后台统一管理，从环境变量读取
    supports_vision: bool = False  # 是否支持视觉输入
    
    def model_post_init(self, __context):
        """Pydantic V2的初始化后钩子"""
        # 自动检测视觉支持
        if not self.supports_vision:
            self.supports_vision = self._detect_vision_support()
    
    def _detect_vision_support(self) -> bool:
        """检测模型是否支持视觉"""
        vision_models = [
            'gpt-4-vision',
            'gpt-4-turbo',
            'gpt-4o',
            'claude-3-opus',
            'claude-3-sonnet',
            'claude-3-haiku'
        ]
        return any(vm in self.model.lower() for vm in vision_models)


class MemoryConfig(BaseModel):
    """记忆配置"""
    short_term: Dict = field(default_factory=lambda: {
        "enabled": True,
        "window_size": 10
    })
    long_term: Dict = field(default_factory=lambda: {
        "enabled": True,
        "retention_days": 90
    })
    working: Dict = field(default_factory=lambda: {
        "enabled": True
    })


class AgentConfig(BaseModel):
    """Agent配置"""
    model_config = {}  # Pydantic V2配置，必须在类定义开始
    
    id: str
    user_id: str
    name: str
    description: str
    type: Literal["basic", "react", "deep_research"]
    avatar_url: Optional[str] = None
    tags: List[str] = []
    
    # LLM模型配置（避免与Pydantic的model_config冲突）
    llm_config: ModelConfig
    
    # Prompt配置
    system_prompt: str
    prompt_template: Optional[str] = None
    
    # 记忆配置
    memory_config: MemoryConfig = MemoryConfig()
    
    # 知识库绑定
    knowledge_base_ids: List[str] = []
    
    # 工具绑定
    tool_ids: List[str] = []
    
    # ReAct特定配置
    react_config: Optional[Dict] = None
    
    # Research特定配置
    research_config: Optional[Dict] = None
    
    created_at: datetime = field(default_factory=datetime.now)
    updated_at: datetime = field(default_factory=datetime.now)
    
    @validator('name')
    def validate_name(cls, v):
        if len(v) < 2 or len(v) > 100:
            raise ValueError('名称长度必须在2-100之间')
        return v
    
    @validator('description')
    def validate_description(cls, v):
        if len(v) > 500:
            raise ValueError('描述长度不能超过500')
        return v


# ============= 会话和消息模型 =============

@dataclass
class Conversation:
    """会话"""
    id: str
    agent_id: str
    user_id: str
    title: str
    summary: Optional[str] = None
    message_count: int = 0
    created_at: datetime = field(default_factory=datetime.now)
    updated_at: datetime = field(default_factory=datetime.now)


@dataclass
class Message:
    """消息"""
    id: str
    conversation_id: str
    role: Literal["system", "user", "assistant", "function"]
    content: str
    metadata: Dict = field(default_factory=dict)
    created_at: datetime = field(default_factory=datetime.now)


@dataclass
class StreamChunk:
    """流式响应块"""
    delta: str
    finish_reason: Optional[str] = None


# ============= API请求/响应模型 =============

class AgentCreateRequest(BaseModel):
    """Agent创建请求"""
    name: str
    description: str
    type: Literal["basic", "react", "deep_research"]
    avatar_url: Optional[str] = None
    tags: List[str] = []
    llm_config: ModelConfig
    system_prompt: str
    prompt_template: Optional[str] = None
    memory_config: Optional[MemoryConfig] = None
    knowledge_base_ids: List[str] = []
    tool_ids: List[str] = []
    react_config: Optional[Dict] = None
    research_config: Optional[Dict] = None
    
    @validator('name')
    def validate_name(cls, v):
        if len(v) < 2 or len(v) > 100:
            raise ValueError('名称长度必须在2-100之间')
        return v
    
    @validator('description')
    def validate_description(cls, v):
        if len(v) > 500:
            raise ValueError('描述长度不能超过500')
        return v


class AgentUpdateRequest(BaseModel):
    """Agent更新请求"""
    name: Optional[str] = None
    description: Optional[str] = None
    avatar_url: Optional[str] = None
    tags: Optional[List[str]] = None
    llm_config: Optional[ModelConfig] = None
    system_prompt: Optional[str] = None
    prompt_template: Optional[str] = None
    memory_config: Optional[MemoryConfig] = None
    knowledge_base_ids: Optional[List[str]] = None
    tool_ids: Optional[List[str]] = None
    react_config: Optional[Dict] = None
    research_config: Optional[Dict] = None


class ConversationCreateRequest(BaseModel):
    """会话创建请求"""
    agent_id: str
    title: Optional[str] = "新对话"


class MessageSendRequest(BaseModel):
    """消息发送请求"""
    content: str
    stream: bool = True
    attachments: List[Dict] = []

File: desktop/frontend/src/types/agent.ts
/**
 * Agent相关类型定义
 */

export type AgentType = "basic" | "react" | "deep_research";

export type LLMProvider = "openai" | "claude";

export interface ModelConfig {
  provider: LLMProvider;
  model: string;
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  frequency_penalty?: number;
  presence_penalty?: number;
  // API Key 和 base_url 由后台统一管理，客户端不需要配置
  supports_vision?: boolean;
}

export interface MemoryConfig {
  short_term?: {
    enabled: boolean;
    window_size: number;
  };
  long_term?: {
    enabled: boolean;
    retention_days: number;
  };
  working?: {
    enabled: boolean;
  };
}

export interface AgentConfig {
  id: string;
  user_id: string;
  name: string;
  description: string;
  type: AgentType;
  avatar_url?: string;
  tags: string[];
  llm_config: ModelConfig;
  system_prompt: string;
  prompt_template?: string;
  memory_config: MemoryConfig;
  knowledge_base_ids: string[];
  tool_ids: string[];
  react_config?: {
    max_iterations: number;
  };
  research_config?: {
    complexity_threshold: number;
    max_subtasks: number;
  };
  created_at: string;
  updated_at: string;
}

export interface AgentCreateRequest {
  name: string;
  description: string;
  type: AgentType;
  avatar_url?: string;
  tags?: string[];
  llm_config: ModelConfig;
  system_prompt: string;
  prompt_template?: string;
  memory_config?: MemoryConfig;
  knowledge_base_ids?: string[];
  tool_ids?: string[];
  react_config?: {
    max_iterations: number;
  };
  research_config?: {
    complexity_threshold: number;
    max_subtasks: number;
  };
}

export interface AgentUpdateRequest {
  name?: string;
  description?: string;
  avatar_url?: string;
  tags?: string[];
  llm_config?: ModelConfig;
  system_prompt?: string;
  prompt_template?: string;
  memory_config?: MemoryConfig;
  knowledge_base_ids?: string[];
  tool_ids?: string[];
  react_config?: {
    max_iterations: number;
  };
  research_config?: {
    complexity_threshold: number;
    max_subtasks: number;
  };
}

export interface Conversation {
  id: string;
  agent_id: string;
  user_id: string;
  title: string;
  summary?: string;
  message_count: number;
  created_at: string;
  updated_at: string;
}

export interface Message {
  id: string;
  conversation_id: string;
  role: "system" | "user" | "assistant" | "function";
  content: string;
  metadata?: Record<string, any>;
  created_at: string;
}

export interface MessageSendRequest {
  content: string;
  stream?: boolean;
  attachments?: Array<{
    type: "image" | "file";
    url?: string;
    path?: string;
  }>;
}

export interface KnowledgeBase {
  id: string;
  user_id: string;
  name: string;
  description: string;
  embedding_model: string;
  chunk_size: number;
  chunk_overlap: number;
  document_count: number;
  vector_count: number;
  created_at: string;
  updated_at: string;
}

export interface Document {
  id: string;
  knowledge_base_id: string;
  name: string;
  file_path: string;
  file_type: string;
  file_size: number;
  status: "pending" | "processing" | "completed" | "failed";
  chunk_count: number;
  error_message?: string;
  created_at: string;
  processed_at?: string;
}

export interface SearchResult {
  content: string;
  similarity: number;
  document_name: string;
  metadata: Record<string, any>;
}

export interface Tool {
  id: string;
  name: string;
  description: string;
  type: "workflow" | "mcp" | "http" | "system" | "builtin";
  category: string;
  parameters_schema: Record<string, any>;
  config: Record<string, any>;
  approval_policy: "auto" | "required" | "manual";
  allowed_agents: string[];
  is_enabled: boolean;
  created_at: string;
}

export interface ToolCall {
  id: string;
  tool_id: string;
  agent_id: string;
  conversation_id: string;
  input_params: Record<string, any>;
  output_result?: Record<string, any>;
  status: "pending" | "approved" | "rejected" | "executing" | "completed" | "failed";
  execution_time_ms?: number;
  created_at: string;
  completed_at?: string;
}

export interface AgentTemplate {
  id: string;
  user_id?: string;
  name: string;
  description: string;
  category: string;
  type: AgentType;
  tags: string[];
  icon: string;
  is_system: boolean;
  is_public: boolean;
  config: {
    system_prompt: string;
    llm_config: Partial<ModelConfig>;
    memory_config?: MemoryConfig;
    react_config?: {
      max_iterations: number;
    };
    research_config?: {
      complexity_threshold: number;
      max_subtasks: number;
    };
  };
  usage_count: number;
  created_at: string;
  updated_at: string;
}

export interface AgentTemplateCreateRequest {
  name: string;
  description: string;
  category: string;
  type: AgentType;
  tags?: string[];
  icon?: string;
  is_public?: boolean;
  config: AgentTemplate["config"];
}

export interface AgentTemplateQueryParams {
  category?: string;
  type?: AgentType;
  tags?: string[];
  search?: string;
  is_system?: boolean;
  is_public?: boolean;
  page?: number;
  page_size?: number;
}
